File: ui/src/api/locations.ts
import { loggedInvoke } from '@/utils/logged-invoke';

/** A physical location profile with address and local configuration. */
export interface LocationProfile {
  id: string;
  name: string;
  address: string;
  tax_id: string;
  currency: string;
  timezone: string;
  is_primary: boolean;
  created_at: string;
  updated_at: string;
}

/** Arguments for creating a new location profile. */
export interface CreateLocationArgs {
  id: string;
  name: string;
  address?: string;
  tax_id?: string;
  currency?: string;
  timezone?: string;
}

/** Arguments for updating an existing location profile. */
export interface UpdateLocationArgs {
  id: string;
  name: string;
  address: string;
  tax_id: string;
  currency: string;
  timezone: string;
}

/** List all location profiles for the session's tenant (scoped — ADR #7). */
export const listLocationsScoped = (sessionToken: string): Promise<LocationProfile[]> =>
  loggedInvoke<LocationProfile[]>('list_locations_scoped', { sessionToken });

/** Get a single location profile by its identifier (scoped — ADR #7). */
export const getLocationProfileScoped = (
  sessionToken: string,
  id: string,
): Promise<LocationProfile | null> =>
  loggedInvoke<LocationProfile | null>('get_location_profile_scoped', { sessionToken, id });

/** Get the primary location profile (scoped — ADR #7). */
export const getPrimaryLocationScoped = (sessionToken: string): Promise<LocationProfile | null> =>
  loggedInvoke<LocationProfile | null>('get_primary_location_scoped', { sessionToken });

/** Create a new location profile (scoped — ADR #7). */
export const createLocationProfileScoped = (
  sessionToken: string,
  args: CreateLocationArgs,
): Promise<LocationProfile> =>
  loggedInvoke<LocationProfile>('create_location_profile_scoped', { sessionToken, args });

/** Update an existing location profile (scoped — ADR #7). */
export const updateLocationProfileScoped = (
  sessionToken: string,
  args: UpdateLocationArgs,
): Promise<LocationProfile> =>
  loggedInvoke<LocationProfile>('update_location_profile_scoped', { sessionToken, args });

/** Set a location as the primary location (scoped — ADR #7). */
export const setPrimaryLocationScoped = (sessionToken: string, id: string): Promise<LocationProfile> =>
  loggedInvoke<LocationProfile>('set_primary_location_scoped', { sessionToken, id });

/** Delete a location profile by its identifier (scoped — ADR #7). */
export const deleteLocationProfileScoped = (sessionToken: string, id: string): Promise<void> =>
  loggedInvoke<void>('delete_location_profile_scoped', { sessionToken, id });

/** Read a location's ticket prefix (scoped — ADR #7; W7-A UI half).
 *
 *  null = no prefix set — tickets number bare (`#{n}`); a prefix renders
 *  verbatim as `#{prefix}{n}` (render ruling: no invented padding). The
 *  backend normalizes (trim + uppercase) before storing. */
export const getLocationTicketPrefixScoped = (
  sessionToken: string,
  id: string,
): Promise<string | null> =>
  loggedInvoke<string | null>('get_location_ticket_prefix_scoped', { sessionToken, id });

/** Set — or, with an empty string, CLEAR — a location's ticket prefix
 *  (scoped — ADR #7; W7-A UI half).
 *
 *  THE FROZEN-AT-STAMPING RULE (D16): the prefix is copied onto each ticket
 *  when it is stamped, so changing it here affects ONLY future tickets;
 *  existing tickets keep the prefix they were stamped with. The backend
 *  echoes the normalized value (trim + uppercase); empty normalizes to
 *  null (no prefix). */
export const setLocationTicketPrefixScoped = (
  sessionToken: string,
  id: string,
  prefix: string,
): Promise<string | null> =>
  loggedInvoke<string | null>('set_location_ticket_prefix_scoped', {
    sessionToken,
    id,
    prefix,
  });

File: apps/desktop-client/src/commands/locations.rs
//! Tauri commands for location-profile CRUD.
//!
//! Wire contract: each command talks to the `oz_core::Store` facade via the shared
//! `AppState` database connection. The old store-profile command names
//! were retired once every client migrated to the canonical commands
//! (todo-global-saas-1.md, rename slice 1c/1d).
//!
//! Wave E / E7: the bodies now live in the headless `oz_bridge::locations` module.
//! Each `#[tauri::command]` below keeps its exact name, parameter list, attributes
//! and `Result<_, AppError>` wire contract; it builds a `BridgeCtx` from `AppState` and
//! delegates, preserving resolve/gate order, the ADR #47 location-resource
//! checks, the C1.2 quota gate and the drop-then-delete-db ordering. The DTOs
//! moved with the bodies and are re-exported so `use super::*` in
//! `locations_tests.rs` still resolves them.

use tauri::State;

use crate::error::AppError;
use crate::state::AppState;

pub use oz_bridge::locations::{CreateLocationArgs, LocationProfileDto, UpdateLocationArgs};

/// List location profiles for the session's tenant (ADR #7).
#[tauri::command]
pub async fn list_locations_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<LocationProfileDto>, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::list_locations_scoped(&ctx, &session_token)
        .await
        .map_err(Into::into)
}

/// Get a location profile for the session's tenant (ADR #7).
#[tauri::command]
pub async fn get_location_profile_scoped(
    id: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Option<LocationProfileDto>, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::get_location_profile_scoped(&ctx, &session_token, &id)
        .await
        .map_err(Into::into)
}

/// Get the primary location for the session's tenant (ADR #7).
#[tauri::command]
pub async fn get_primary_location_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Option<LocationProfileDto>, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::get_primary_location_scoped(&ctx, &session_token)
        .await
        .map_err(Into::into)
}

/// Create a location profile for the session's tenant (ADR #7).
#[tauri::command]
pub async fn create_location_profile_scoped(
    args: CreateLocationArgs,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<LocationProfileDto, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::create_location_profile_scoped(&ctx, &session_token, &args)
        .await
        .map_err(Into::into)
}

/// Update a location profile for the session's tenant (ADR #7).
#[tauri::command]
pub async fn update_location_profile_scoped(
    args: UpdateLocationArgs,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<LocationProfileDto, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::update_location_profile_scoped(&ctx, &session_token, &args)
        .await
        .map_err(Into::into)
}

/// Set a location as primary for the session's tenant (ADR #7).
#[tauri::command]
pub async fn set_primary_location_scoped(
    id: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<LocationProfileDto, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::set_primary_location_scoped(&ctx, &session_token, &id)
        .await
        .map_err(Into::into)
}

/// Delete a location profile for the session's tenant (ADR #7).
#[tauri::command]
pub async fn delete_location_profile_scoped(
    id: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::delete_location_profile_scoped(&ctx, &session_token, &id)
        .await
        .map_err(Into::into)
}

// ── KDS ticket prefix (W2-A consumer slice) ────────────────────
//
// Two commands over the landed core pair (`location_ticket_prefix` /
// `set_location_ticket_prefix`). The field deliberately does NOT ride on
// `LocationProfileDto`: the core slice chose the dedicated-getter shape (same
// row as `legal_entity_id`), so the profile DTO stays as wide as the table the
// settings screen reads and a prefix edit cannot fail an unrelated save.

/// Read one location's KDS ticket prefix for the session's tenant.
///
/// `None` means "no prefix" — the same sentinel the column stores as `''`.
/// There is no fallback to the legal entity's statutory fiscal prefix, by
/// design: a fiscal re-registration must not retitle kitchen tickets.
#[tauri::command]
pub async fn get_location_ticket_prefix_scoped(
    id: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Option<String>, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::get_location_ticket_prefix_scoped(&ctx, &session_token, &id)
        .await
        .map_err(Into::into)
}

/// Set (or clear) one location's KDS ticket prefix.
///
/// Returns the value AS STORED, after the core's normalization (trim +
/// ASCII-uppercase), so the caller echoes what will be enforced rather than
/// what was typed — `" kds-a "` comes back as `Some("KDS-A")`. An empty or
/// whitespace-only prefix clears it. Unknown ids are `NotFound`; a prefix
/// already used by another location of this tenant surfaces the partial
/// unique index's violation, unsmoothed.
#[tauri::command]
pub async fn set_location_ticket_prefix_scoped(
    id: String,
    prefix: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Option<String>, AppError> {
    let ctx = state.bridge_ctx();
    oz_bridge::locations::set_location_ticket_prefix_scoped(&ctx, &session_token, &id, &prefix)
        .await
        .map_err(Into::into)
}
